File: go/conformance/report.go
package conformance

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

func details(r VectorResult) string {
	if r.Status == "not-applicable" || r.Status == "error" {
		return r.Error
	}
	var out []string
	for _, s := range r.Steps {
		for _, f := range s.Failures {
			out = append(out, s.StepID+": "+f)
		}
	}
	return strings.Join(out, "; ")
}

var xmlEscaper = strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;", "\"", "&quot;")

// Format renders a summary as json, markdown or junit with the same shapes as the TypeScript runner (REQ-CONF-6).
func Format(s Summary, format, target string) ([]byte, error) {
	switch format {
	case "json":
		type report struct {
			Target      string `json:"target,omitempty"`
			GeneratedAt string `json:"generatedAt"`
			Summary
		}
		return json.MarshalIndent(report{Target: target, GeneratedAt: time.Now().UTC().Format(time.RFC3339), Summary: s}, "", "  ")
	case "markdown":
		var b strings.Builder
		b.WriteString("# anyonce conformance report\n\n")
		if target != "" {
			fmt.Fprintf(&b, "Target: %s\n\n", target)
		}
		fmt.Fprintf(&b, "%d passed, %d failed, %d not applicable, %d errored\n\n", s.Passed, s.Failed, s.NotApplicable, s.Errored)
		b.WriteString("| Vector | Tier | Status | Details |\n|---|---|---|---|\n")
		for _, r := range s.Results {
			fmt.Fprintf(&b, "| %s | %s | %s | %s |\n", r.ID, r.Tier, r.Status, strings.ReplaceAll(details(r), "|", "\\|"))
		}
		return []byte(b.String()), nil
	case "junit":
		var b strings.Builder
		b.WriteString("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n")
		fmt.Fprintf(&b, "<testsuite name=\"anyonce-conformance\" tests=\"%d\" failures=\"%d\" errors=\"%d\" skipped=\"%d\">\n", len(s.Results), s.Failed, s.Errored, s.NotApplicable)
		for _, r := range s.Results {
			open := fmt.Sprintf("<testcase classname=\"%s\" name=\"%s\"", r.Tier, xmlEscaper.Replace(r.ID))
			message := xmlEscaper.Replace(details(r))
			switch r.Status {
			case "pass":
				fmt.Fprintf(&b, "  %s/>\n", open)
			case "fail":
				fmt.Fprintf(&b, "  %s><failure message=\"%s\"/></testcase>\n", open, message)
			case "not-applicable":
				fmt.Fprintf(&b, "  %s><skipped message=\"%s\"/></testcase>\n", open, message)
			default:
				fmt.Fprintf(&b, "  %s><error message=\"%s\"/></testcase>\n", open, message)
			}
		}
		b.WriteString("</testsuite>\n")
		return []byte(b.String()), nil
	default:
		return nil, fmt.Errorf("conformance: unknown report format %q", format)
	}
}

File: packages/conformance/src/report.ts
import type { RunSummary, VectorResult } from './types';

export type ReportFormat = 'json' | 'markdown' | 'junit';

export interface ReportMeta {
  target?: string;
  generatedAt?: string;
}

function details(result: VectorResult): string {
  if (result.status === 'not-applicable' || result.status === 'error') return result.error ?? '';
  const out: string[] = [];
  for (const step of result.steps)
    for (const failure of step.failures) out.push(`${step.stepId}: ${failure}`);
  return out.join('; ');
}

function escapeXml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

function markdown(summary: RunSummary, meta: ReportMeta): string {
  const lines = ['# anyonce conformance report', ''];
  if (meta.target !== undefined) lines.push(`Target: ${meta.target}`, '');
  lines.push(
    `${summary.passed} passed, ${summary.failed} failed, ${summary.notApplicable} not applicable, ${summary.errored} errored`,
    '',
    '| Vector | Tier | Status | Details |',
    '|---|---|---|---|',
  );
  for (const result of summary.results) {
    lines.push(
      `| ${result.id} | ${result.tier} | ${result.status} | ${details(result).replace(/\|/g, '\\|')} |`,
    );
  }
  lines.push('');
  return lines.join('\n');
}

function testcase(result: VectorResult): string {
  const open = `<testcase classname="${result.tier}" name="${escapeXml(result.id)}"`;
  const message = escapeXml(details(result));
  switch (result.status) {
    case 'pass':
      return `${open}/>`;
    case 'fail':
      return `${open}><failure message="${message}"/></testcase>`;
    case 'not-applicable':
      return `${open}><skipped message="${message}"/></testcase>`;
    case 'error':
      return `${open}><error message="${message}"/></testcase>`;
  }
}

function junit(summary: RunSummary): string {
  const cases = summary.results.map(testcase);
  return [
    '<?xml version="1.0" encoding="UTF-8"?>',
    `<testsuite name="anyonce-conformance" tests="${summary.results.length}" failures="${summary.failed}" errors="${summary.errored}" skipped="${summary.notApplicable}">`,
    ...cases.map((c) => `  ${c}`),
    '</testsuite>',
    '',
  ].join('\n');
}

/** REQ-CONF-5: the three report formats share one summary shape. */
export function formatReport(
  summary: RunSummary,
  format: ReportFormat,
  meta: ReportMeta = {},
): string {
  switch (format) {
    case 'json':
      return JSON.stringify({ ...meta, ...summary }, null, 2);
    case 'markdown':
      return markdown(summary, meta);
    case 'junit':
      return junit(summary);
  }
}
